Send media events to COGS

chetbox opened the pull request.

chetbox commented on src/AudioPlayer.ts:
```diff
@@ -227,6 +238,25 @@ export default class AudioPlayer {
+        ...clipPlayer,
+        activeClips: Object.fromEntries(
+          Object.entries(clipPlayer.activeClips).map(([soundIdStr, clip]) => {
+            if (clip.state !== ActiveAudioClipState.Stopping) {
```
Yes that's right. Something has already stopped this clip and it's fading out so we want to make sure you don't enable `loop` which will make it start again!

tomdixon commented on src/VideoPlayer.ts:
```diff
@@ -0,0 +1,319 @@
+
+    this.updateVideoClipPlayer(this.activeClipPath, (clipPlayer) => {
+      if (clipPlayer.videoElement) {
+        clipPlayer.videoElement.volume = volume;
```
Global volume not taken into account here?

chetbox replied:
Good find! Fixed.

tomdixon reviewed (changes requested):
A few comments to be addressed. Also found the following in my testing:

**1. Clip default states not sent to COGS when MM connects**
So if I have a clip playing on a master and I think disconnect it and re-connect it, it doesn't send up the default state for any clips which means COGS remains thinking the clip is playing forever. Affects audio & video.

**2. Video elements missing**
I found if I added a "Video play" action whilst monitoring the Media Master DOM then I'd see the video element be added. If I then remove that action it is removed again. However from then on the video element won't be re-added for that particular clip even if it is referenced again in a "Video play" action.

chetbox commented on src/VideoPlayer.ts:
```diff
@@ -0,0 +1,319 @@
+  setGlobalVolume(volume: number): void {
+    // Adjust volume of all videos taking into account previous global volume
+    Object.values(this.videoClipPlayers).forEach((clipPlayer) => {
+      clipPlayer.videoElement.volume = (clipPlayer.videoElement.volume / this.globalVolume) * volume;
```
Good point! Fixed a part of point 3 on https://github.com/clockwork-dog/cogs-client-lib/pull/29#pullrequestreview-861202732

chetbox commented:
> 3. Global volume
A few things in here relating to global volume. I wonder if we need to separately store the global volume for the master, along with the current playing clip's requested volume. If then either the global volume or playing clip's volumes are altered whilst the clip is playing we re-compute the new volume for the video element at that point and apply it.

This is a good idea and we probably don't want to be subject to floating point rounding errors. This extra bit of state makes it easier to understand IMO. Done. 👍🏼

chetbox commented:
> 2. Video elements missing
I found if I added a "Video play" action whilst monitoring the Media Master DOM then I'd see the video element be added. If I then remove that action it is removed again. However from then on the video element won't be re-added for that particular clip even if it is referenced again in a "Video play" action.

Good spot. There was also a smoking gun here when a video clip was updated. Fixed.

chetbox commented:
> 1. Clip default states not sent to COGS when MM connects
So if I have a clip playing on a master and I think disconnect it and re-connect it, it doesn't send up the default state for any clips which means COGS remains thinking the clip is playing forever. Affects audio & video.

I've added a new client-to-COGS message called `allMediaClipStates` which notifies COGS of what's playing when the client connects to COGS. In the normal case, this list is empty. COGS can assume that any file that is _not_  in this list is not playing on that output. This seems like the neatest and most robust solution to me.

tomdixon commented:
Bug: Update loop whilst audio playing, it doesn't loop.

tomdixon reviewed (approved):
Time for a new lib!

tomdixon commented on src/AudioPlayer.ts:
```diff
@@ -73,36 +101,55 @@ export default class AudioPlayer {
+      console.log('newSoundIds', newSoundIds);
 
       [...pausedSoundIds, ...newSoundIds].forEach((soundId) => {
+        console.log('looping', JSON.stringify(soundId), loop, clipPlayer.player);
```
Debugging and can be removed now?

chetbox replied:
Done

tomdixon commented on src/AudioPlayer.ts:
```diff
@@ -73,36 +101,55 @@ export default class AudioPlayer {
-            ];
+      const newSoundIds = pausedSoundIds.length > 0 ? [] : [clipPlayer.player.play()];
+
+      console.log('newSoundIds', newSoundIds);
```
Debugging and can be removed now?

tomdixon commented on src/AudioPlayer.ts:
```diff
@@ -73,36 +101,55 @@ export default class AudioPlayer {
+        clipPlayer.player.once(
+          'stop',
+          () => {
+            console.log('stop. loop:', !clipPlayer.activeClips[soundId]?.loop);
```
Debugging and can be removed now?

chetbox replied:
Done

tomdixon commented on src/AudioPlayer.ts:
```diff
@@ -73,36 +101,55 @@ export default class AudioPlayer {
           volume,
         };
 
+        setInterval(() => {
```
Debugging and can be removed now?

chetbox replied:
Done

chetbox commented on src/AudioPlayer.ts:
```diff
@@ -73,36 +101,55 @@ export default class AudioPlayer {
+        clipPlayer.player.on(
+          'end',
+          () => {
+            console.log('end. loop:', !clipPlayer.activeClips[soundId]?.loop);
```
Done
